### PR TITLE
Fix category thumbnail images

### LINKED ISSUE #30520: Categories image - harmonization of 2 images
# User story:

Make categories great again.
As a merchant, I want to be able to edit the different categories of images (cover, thumbnail) and using the image set in the back-office.

# Constraint:

→ No BC

The image provided by the back doesn’t send the right data. The file between Category cover image & Category thumbnail are the same.

# Current situation:

* We have two images available to be set to a category in BO. Only one of them is used, one is not available at all and accessing them is a mess.
	* *Category cover image*
		* Currently used for category cover in the header of the category and subcategory thumbnail in the parent category.
		* Accessible by *$category.image.bySize.category_default*
	* *Category thumbnail*
		* Not used in classic theme right now at all.
		* Accessible by *$category.image.bySize.small_default*
	* Image array with *bySize* should contain the same image, but in different sizes, like product images. Not magically a different image for each size. 🙂
* Another problem. In *BO > Image settings*, where we select for which type of objects we want the thumbnail to be generated for. 
	* All other types of objects (product, manufacturer) have only 1 type of image. Either a single one (manufacturer, supplier, store) OR in case of products, multiple images, but serving the same purpose, having the same sizes.
	* But for categories, we have 2 images with different usage (to have in the same time a cover and the thumbnail with a different images).
	* How to solve this?
		* A) Either generate two images in all thumbnail definitions, which will leave some thumbnails unused. So for example, size *‘cover_large’* will be generated for *$category.thumbnail*, but not used at all.
		* B) Or extend upon current settings in image sizes and instead of one “Category” type,  add “Category cover”, “Category thumbnail”. So we can configure sizes for each type of image separately, so we don’t waste space.

# Specs
In FO *$category* object, we need to have two images available as a standardized array with all the sizes. It should for example be *$category.thumbnail* and *$category.cover*.

All of these will be an array as we know from current *$category.image* or product images, but with correct data.

The same should be standardized for all objects in FO, manufacturers for example. Which right now provide only a single URL in *$manufacturer.image*, instead of array with all sizes.

One question to decide. On product page, we have *$product.images* array with the images.
Do we want to have *$category.thumbnail* and *$category.cover* separately? Or put it to a new array called *$category.images* with children thumbnail and cover?

_____________________________________________

# @Hlavtox specs proposal on 22.5.2023
# Basic things to resolve
1. We need to be able to have this data in front office.
2. We need to review a system of how these images are stored on the filesystem.
3. We need to properly retrieve links for these images.
4. We need to be access these images by the rewritten URL.
5. We need ImageRetriever to properly generate and retrieve thumbnails for these images.
6. We need backoffice manual regenerate AND regenerate after upload to do it properly also.
7. We need to be able to define sizing for these images, possibly for other image types.

# 1. We need to be able to have this data in front office. ✅
https://github.com/PrestaShop/PrestaShop/pull/32653

# 2. We need to review a system of how these images are stored on the filesystem. 🟡 
- Original image for cover - `/img/c/1.jpg`, working
- Original image for thumbnail - `/img/c/1_thumb.jpg`, working
- Resized image(s) for cover - `/img/c/1-category_default.jpg`, working
- Resized image(s) for thumbnail - `/img/c/1_thumb-category_default.jpg`. Currently broken, backoffice is generating it as `/img/c/3-small_default.jpg` overwriting the cover image miniature.

# 3. We need to properly retrieve links for these images. ✅
Link class must be able to provide links to these images properly. Works, tested.

# 4. We need to be access these images by the rewritten URL. 🔴 
Rewrites do not work.
`http://localhost/dev/c/4-category_default/men.jpg` works
`http://localhost/dev/c/4_thumb-category_default/men.jpg` does not

There are rewrites only for the basic image.
```
RewriteCond %{HTTP_HOST} ^localhost$
RewriteRule ^c/([\d]+)(\-[\.*\w-]*)/.+(\.(?:jpe?g|webp|png|avif))$ %{ENV:REWRITEBASE}img/c/$1$2$3 [L]
RewriteCond %{HTTP_HOST} ^localhost$
RewriteRule ^c/([a-zA-Z_-]+)(-[\d]+)?/.+(\.(?:jpe?g|webp|png|avif))$ %{ENV:REWRITEBASE}img/c/$1$2$3 [L]
```

Manufacturers don't use any rewrites.
`http://localhost/dev/img/m/1-small_default.webp`
Maybe we can do the same for category images? Do they need rewriting?

# 5. We need ImageRetriever to properly generate and retrieve thumbnails for these images. ✅
After a small change (allowing a string as image id) and passing the ID image with _thumb suffix, ImageRetriever retrieves a proper set of images, in all formats configured.

Snímek obrazovky 2023-05-20 165004

# 6. We need backoffice to properly generate these thumbnails. 🔴 
Backoffice is currently generating it as `/img/c/3-small_default.jpg` instead of `/img/c/3_thumb-small_default.jpg`, overwriting the cover image miniature.

# 7. We need to be able to define sizing for these images, possibly for other image types. 🔴 
Miniatures that should be generated are stored in ps_image_type table. This table has a problem, it only says that the current size is for categories, but not anything else. There is no way to define additional details.

We can currently get a way by generating both cover and thumbnail in all sizes defined for categories. This would bring SOME additional storage needed, but category counts are in hundreds on most stores. This is how it currently works.

Current table structure looks like this:
| id_image_type	 | name	 | width	 | height	 | products	 | categories	 | manufacturers	 | suppliers	 | stores |
| ---                           | ---          | ---     | ---     | ---      | ---          | ---             | ---         | ---    |
| 1	                          | cart_default	 | 125	 | 125    | 1       | 0	   | 0	           | 0	              | 0 |
| 2	                          | small_default	 | 98	         | 98	    | 1       | 1	   | 1	           | 1	               | 0 |

My proposal is to change it like this, making each size it's own line with it's own purpose. The table would not be much bigger because the image types hardly interfere with each other. Also, it would allow to use this system by custom modules for their images. The change would be done internally and `ImageType::getImagesTypes('products')`; would return the same data.

| id_image_type	 | name	 | width	 | height	 | purpose	 | 
| ---                           | ---          | ---          | ---          | ---                 | 
| 1	                        | cart_default  | 125	 | 125          |    products     |
| 2	                     | cover_default	 | 1000	  | 100	   | category_cover |
| 3	                     | cover_default_2x	 | 2000	  | 200	   | category_cover |
| 4	                     | small_default	 | 98	  | 98	   | category_thumbnail |
| 5	                     | small_default	 | 98	  | 98	   | blog_post_thumbnail |

We can further extend this by adding a column for non-deletable thumbnails for backoffice purposes, which would allow us to get rid of the old thumbnail images generated somewhere. Also, we can add a new column with a desired thumbnail fitment, so we can implement https://github.com/PrestaShop/PrestaShop/issues/26158 properly.

# 8 Acceptance Test
https://github.com/PrestaShop/PrestaShop/blob/d672ad4ccc467722d13746e25fffc95a86e5e3b3/tests/UI/campaigns/functional/BO/08_design/06_imageSettings/11_imageGenerationOnCreation.ts#L301-L302

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | Fixes category cover vs thumbnail conflict, renders both properly in front office. See more below
| Type?             | bug fix
| Category?         | CO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | read below.
| UI Tests          | https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/10903585728 https://github.com/florine2623/testing_pr/actions/runs/11009547904 ✅ 
| Fixed issue or discussion?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/30520
| Related PRs       | https://github.com/PrestaShop/autoupgrade/pull/1060
| Sponsor company   | 

### What is it about
- Category has two images. One is stored in `ID.jpg`, the second in `ID_thumb.jpg`.
- The logic is currently broken and when you want to get category image in front office, you randomly get a different image in small_default array.

### What this fixes
- You can now get both images in all dimensions via functions I prepared earlier in `CategoryLazyArray`. Just by calling `$category->thumbnail`, `$category->cover`.
- [BC break] No image is now returned by default, if the image does not exist. Theme can decide what to do. Now, it didn't have a chance to know if the image was returned or the placeholder was returned.
- You can now manage both images in backoffice separately.
  - They are not overwriting each other anymore.
  - Changing cover doesn't change thumbnail anymore.
  - You can delete both images separately.
  - Thumbnails are correctly deleted in all formats.
- `.htaccess` rules have been added to rewrite links to these images.
- Both cover and thumbnail is correctly generated when using regeneration function in backoffice.
- Simplifies and speeds up thumbnail deletion, because it uses one regex to check if it's a thumbnail.

### Theme change required
- Changing `$category->image` to `$category->thumbnail` and `$category->cover`.
- If image does not exist, optionally render `$urls.no_picture_image`.
- See example in this PR - https://github.com/PrestaShop/classic-theme/pull/154

----------

### Autoupgrade PR logic
```php
/*
 * Pre v9 worked like this:
 * There could be 123.jpg, a cover image.
 * There could be 123_thumb.jpg, a category miniature.
 * 
 * If 123_thumb.jpg existed, it took it and generated category thumbnails from it.
 * But, overwriting thumbnails of 123.jpg.
 *
 * So you suddenly got one thumbnail that was from a different image.
 * Even worse, frontoffice and backoffice generated it as a different thumbnail.
 */
$formattedNameSmall = ImageType::getFormattedName('small');
$formattedNameMedium = ImageType::getFormattedName('medium');

foreach ($categoryIds as $categoryId) {
  /*
   * Define paths to original files.
   * Define paths to possible broken thumbnails.
   */
  $categoryCoverPath = _PS_CAT_IMG_DIR_ . $categoryId . '.jpg';
  $categoryMiniaturePath = _PS_CAT_IMG_DIR_ . $categoryId . '_thumb.jpg';
  $categorySmallPath =_PS_CAT_IMG_DIR_ . $categoryId . '-' . $formattedNameSmall . '.jpg';
  $categoryMediumPath = _PS_CAT_IMG_DIR_ . $categoryId . '-' . $formattedNameMedium . '.jpg';

  // Check what images exist
  $hasCover = file_exists($categoryCoverPath);
  $hasMiniature = file_exists($categoryMiniaturePath);
  
  /*
   * If both exist, generated thumbnails can be wrong.
   * We delete them and we are finished.
   */
  if ($hasCover && $hasMiniature) {
    if (file_exists($categorySmallPath)) {
      unlink($categorySmallPath);
    }
    if (file_exists($categoryMediumPath)) {
      unlink($categoryMediumPath);
    }
  }

  /*
   * If it has a cover but no thumbnail, we don't want people
   * to end up with no thumbnail. We will copy the cover
   * as the thumbnail. 
   */
  if ($hasCover && !$hasMiniature) {
    copy($categoryCoverPath, $categoryMiniaturePath);
  }

  /*
   * If it has a thumbnail but no cover, we will delete thumbnails,
   * because they have a wrong name and we don't want them to make
   * issues.
   */
  if (!$hasCover && $hasMiniature) {
    if (file_exists($categorySmallPath)) {
      unlink($categorySmallPath);
    }
    if (file_exists($categoryMediumPath)) {
      unlink($categoryMediumPath);
    }
  }
}
```

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `BC breaks` should be `yes` or `no`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#bc-breaks))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
### About linked issues
Please consider opening an issue before submitting a Pull Request:
* If it's a bug fix, it helps maintainers verify that the bug is effectively due to a defect in the code, and that it hasn't been fixed already.
* It can help trigger a discussion about the best implementation path before a single line of code is written.
* It may lead the Core Product team to mark that issue as a priority, further attracting the maintainers' attention.

(Note: this is an automated message, but answering it will reach a real human)

File: src/Adapter/Presenter/Category/CategoryPresenter.php
Comment (kpodemski) [inline src/Adapter/Presenter/Category/CategoryPresenter.php]:
```suggestion
```

:D

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `BC breaks` should be `yes` or `no`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#bc-breaks))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
### About linked issues
Please consider opening an issue before submitting a Pull Request:
* If it's a bug fix, it helps maintainers verify that the bug is effectively due to a defect in the code, and that it hasn't been fixed already.
* It can help trigger a discussion about the best implementation path before a single line of code is written.
* It may lead the Core Product team to mark that issue as a priority, further attracting the maintainers' attention.

(Note: this is an automated message, but answering it will reach a real human)

**Review by ShaiMagal — DISMISSED** [review]

**Review by tleon — DISMISSED** [review]

**Comment by florine2623** [comment]
Hello @PrestaShop/qa-automation team !

Could you check the automated tests, they are broken because of this PR.
Could you confirm ?

Thanks!

**Comment by Hlavtox** [comment]
@nesrineabdmouleh @florine2623 I will check if I can adapt the template in a way, so the tests pass. 👍

**Review by ShaiMagal — DISMISSED** [review]

**Comment by Hlavtox** [comment]
@tleon @kpodemski I put back one method to reduce a BC break, can you reapprove please? :-)

**Comment by Hlavtox** [comment]
Waiting for:
- https://github.com/PrestaShop/classic-theme/pull/154
- https://github.com/PrestaShop/hummingbird/pull/654

**Comment by Hlavtox** [comment]
@florine2623 @nesrineabdmouleh @kpodemski Tests are now full green! :-) https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/10903585728 ✅

**Review by florine2623 — COMMENTED** [review]
Hello @Hlavtox ,

Tested the following on `classic` and `hummingbird`.

No image is now returned by default, if the image does not exist. ✅ 

https://github.com/user-attachments/assets/876b0911-dd8b-4a93-898e-839fbd053e6f

They are not overwriting each other anymore. ✅ 
All the images types category thumbnail and category cover image have their own place on the file system.

Changing/deleting cover doesn't change thumbnail anymore. :x:
When I add a category thumbnail or a category cover image and the format is not correct, it will delete the image I have already saved.

https://github.com/user-attachments/assets/9ce887c8-fda9-4b92-9010-79be03ec4549

You can delete both images separately. :x:
I have an exception when I delete my Category thumbnail :

Screenshot 2024-09-23 at 15 07 16

Thumbnails are correctly deleted in all formats. :x:
They are deleted only on for Category cover image as I have an exception when I try to delete my Category thumbnail. This can't be tested properly.

https://github.com/user-attachments/assets/9b83665b-ffee-4328-a43c-1e09b488d5d9

Both cover and thumbnail is correctly generated when using regeneration function in backoffice ✅ 

Screenshot 2024-09-23 at 15 10 24

Tested this on classic and hummingbird : `Changing $category->image to $category->thumbnail and $category->cover.`
It works as expected ✅ 


Could you check those points in red ? ^^
Thank you :)

**Comment by Hlavtox** [comment]
@florine2623 I can't reproduce it for some reason, and the error really makes no sense, because the same exact code `getCommandBus` is used on many places in the controller. 🤔 

Could you try it again and make sure the cache is deleted? Weird.

Could you help please @jolelievre or @matthieu-rolland? 🙏

**Comment by jolelievre** [comment]
Hi @Hlavtox 

it may be related to the refacto we are doing recently, all the BO controllers are refactored to be based on `PrestashopAdminController` and not `FrameworkBundleAdminController` anymore, as such the protected helper methods are not the same.

The PR for this controller was merged a few days ago https://github.com/PrestaShop/PrestaShop/pull/36883 so you need to rebase you PR and maybe update any controller that has been refactored. For example, the `getCommandBus` no longer exists, instead you can use the `dispatchCommand` method directly

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php]: Why would you handle this separately, the security is supposed to be handled by the framework automatically everywhere Why do you use a custom CSRF token here?

File: src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php]: Why do you use custom CSRF token?

File: src/PrestaShopBundle/Resources/config/services/adapter/category.yml
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/config/services/adapter/category.yml]:
```suggestion
  PrestaShop\PrestaShop\Adapter\Category\CommandHandler\DeleteCategoryThumbnailImageHandler:
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php
Comment (Hlavtox) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php]: Because category cover deletion mechanism uses it also.

File: src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php
Comment (Hlavtox) [reply src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php]: Because category cover deletion mechanism uses it also.

**Comment by Hlavtox** [comment]
Hi @florine2623, can you try to retest the deletion now?

@jolelievre @nesrineabdmouleh Rebased, but tests are failing on the current dev branch. Integration test fails in CI and there is one failure on stock page in UI tests.

File: src/Adapter/Image/Uploader/CategoryThumbnailImageUploader.php
Comment (FabienPapet) [inline src/Adapter/Image/Uploader/CategoryThumbnailImageUploader.php:64]: I'm just putting it here, but it's applicable to all files in this PR and all other PR you submitted. 

Use return types, you'll thank me (and probably yourself) later 😅

**Review by florine2623 — COMMENTED** [review]
Hello @Hlavtox ,

Latest auto tests work are green ✅ 

`Deletion for both images type works as expected` ✅

`Both cover and thumbnail are correctly generated when using regeneration function in backoffice.`
When I regenerate the category images for all formats, I have only 6 images. And before in my previous comment, you can see that I had 16 image files.
Which one is correct ?

Tested yesterday :
Screenshot 2024-09-23 at 15 10 24

Tested today with the latest commit :
Screenshot 2024-09-24 at 15 47 50

`Changing/deleting cover doesn't change thumbnail anymore` ❌
When I add a category thumbnail or a category cover image and the format is not correct, it will delete the image I have already saved.
Steps to reproduce :
- Edit an existing category with a Category cover image
- Add a image with .avif extension > Save
- See that both images are deleted (Category cover image & Category thumbnail)


Waiting for your feedback 2 more approvals before testing the PR again :)

**Comment by Hlavtox** [comment]
@florine2623 The number of thumbnails depends on themes you have installed, and if there are some previously. :-)

> Changing/deleting cover doesn't change thumbnail anymore ❌

I will check this scenario. ;-)

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php
Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php]: But why not use the regular CSRF token in this case? I don't remember any other use case where we need to customize the token this way, so it seems overkill to me 🤔

Comment (Hlavtox) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php]: I can do it Jo, I just did not know, that's all. Still learning. :-) I copied it from the cover mechanism.

Can you please suggest the code changes I need to do to achieve this and simplify this as you say? :-)

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php]: In theory you simply don't need to add any override of the CSRF token, it will be automatically included in the URL And the security system will be able to detect if it's an incorrect one and redirect if needed

File: src/PrestaShopBundle/Resources/config/services/adapter/category.yml
Comment (Hlavtox) [reply src/PrestaShopBundle/Resources/config/services/adapter/category.yml]: Done :-)

File: src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php
Comment (Hlavtox) [reply src/PrestaShopBundle/Form/Admin/Catalog/Category/AbstractCategoryType.php]: Done :-)

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php
Comment (Hlavtox) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php]: Done :-)

**Comment by Hlavtox** [comment]
@florine2623 OK, so I checked it out. Actually, if you upload a faulty image, it doesn't delete anything, but the form images are loaded empty.

So, you open a category with images, delete a name, press save, page reloads with an error next to the name field, but images are missing in the form. F5 and they pop back.

@jolelievre Could you please help me fix it so we can finalize this? 🙏

**Review by tleon — DISMISSED** [review]

**Review by jolelievre — APPROVED** [review]

**Comment by jolelievre** [comment]
@florine2623 I just tested the faulty scenario and I confirm what @Hlavtox said

When you upload an invalid image you get an error and you don't see the image preview anymore, that's because the submitted data from the form overrides the preview data But no image has been removed, if you refresh the page you can see that both images are unmodified

It is a bit inconvenient, but it's a global problem for all our image input types, not relevant to this PR so I don't think it should be blocking the new feature

**Review by kpodemski — APPROVED** [review]

**Comment by Hlavtox** [comment]
Thank you @jolelievre for checking it! :-)

@florine2623 If you agree, I would resolve this for v9 in another PR, so we can reach some milestone with this, it's a big PR. Fixing the form will not be that difficult.

**Review by florine2623 — APPROVED** [review]
Hello @Hlavtox and @jolelievre ,

Ok for me after reading all your comments. Thanks for the details, it is very clear.
It is non blocking for the merge of this PR :)

Thanks!

**Comment by Hlavtox** [comment]
Thank you @florine2623 & @jolelievre!!!

**Pull request merged**
